AppController/djinn.rb

```diff
@@ -4705,72 +4630,160 @@ def place_error_app(app_name, err_msg, language)
     ea.generate(language)
   end
 
-  def start_appengine()
-    @state = "Preparing to run AppEngine apps if needed"
-    Djinn.log_debug("Preparing to run AppEngine apps if needed")
+  # Examine the list of applications to restarts, or the applications that
+  # should be running, and respectively restarts or stats them.
+  def restart_appengine()
+    @state = "Preparing to restart AppEngine apps if needed."
+    Djinn.log_debug(@state)
 
-    if @restored == false
-      # Wait till we can talk to the UserAppServer while we restore.
-      uac = UserAppClient.new(my_node.private_ip, @@secret)
-      Djinn.log_info("Need to restore")
-      begin
-        app_list = uac.get_all_apps()
-      rescue FailedNodeException
-        Djinn.log_warn("Failed to get app listing: retrying.")
-        retry
+    # use a copy of @apps_to_restart here since we delete from it in
+    # setup_appengine_application
+    apps = @apps_to_restart
+    apps.each { |app_name|
+      if !my_node.is_login?  # this node has the new app - don't erase it here
+        Djinn.log_info("Removing old version of app #{app_name}")
+        Djinn.log_run("rm -fv #{PERSISTENT_MOUNT_POINT}/apps/#{app_name}.tar.gz")
       end
-      app_list.each { |app|
+      Djinn.log_info("About to restart app #{app_name}")
+      APPS_LOCK.synchronize {
+        setup_appengine_application(app_name, "restart")
+        maybe_reload_taskqueue_worker(app_name)
+      }
+    }
+  end
+
+  # Check with the UserAppServer the list of application that should be
```

Comment:
Change this comment to something like:

"Compares the list of applications that should be running according to the UserAppServer with the list we have on the load balancer, and marks the missing apps for start during the next cycle."